From 21d4e6c2a2944a42ad92dce691ee705cdbcd1c4b Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 13 Nov 2022 11:09:48 +0100
Subject: [PATCH 0781/1013] evl/wait: share common wait-schedule code

Since we have all the needed information for maintaining a wait list
into the wait channel descriptor embedded into the waitqueue, we may
share the wait-schedule code with other parts of the core,
particularly the mutex support.

Export this common code as __evl_wait_schedule() accordingly, updating
the wait-related tracepoints to operate on wait channels directly,
instead of waitqueues.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/wait.h         |  7 ++++++-
 include/trace/events/evl.h | 36 ++++++++++++++++++------------------
 kernel/evl/mutex.c         | 35 +----------------------------------
 kernel/evl/wait.c          | 20 ++++++++++----------
 4 files changed, 35 insertions(+), 63 deletions(-)

diff --git a/include/evl/wait.h b/include/evl/wait.h
index 7f18c13d5aa0..2c29c2102923 100644
--- a/include/evl/wait.h
+++ b/include/evl/wait.h
@@ -113,7 +113,12 @@ void evl_add_wait_queue_unchecked(struct evl_wait_queue *wq,
 				  ktime_t timeout,
 				  enum evl_tmode timeout_mode);
 
-int evl_wait_schedule(struct evl_wait_queue *wq);
+int __evl_wait_schedule(struct evl_wait_channel *wchan);
+
+static inline int evl_wait_schedule(struct evl_wait_queue *wq)
+{
+	return __evl_wait_schedule(&wq->wchan);
+}
 
 static inline bool evl_wait_active(struct evl_wait_queue *wq)
 {
diff --git a/include/trace/events/evl.h b/include/trace/events/evl.h
index e5894d04405d..8eb7dd714704 100644
--- a/include/trace/events/evl.h
+++ b/include/trace/events/evl.h
@@ -66,19 +66,19 @@ DECLARE_EVENT_CLASS(curr_thread_event,
 		  __entry->state, __entry->info)
 );
 
-DECLARE_EVENT_CLASS(wq_event,
-	TP_PROTO(struct evl_wait_queue *wq),
-	TP_ARGS(wq),
+DECLARE_EVENT_CLASS(wchan_event,
+	TP_PROTO(struct evl_wait_channel *wchan),
+	TP_ARGS(wchan),
 
 	TP_STRUCT__entry(
-		__string(name, wq->wchan.name)
+		__string(name, wchan->name)
 	),
 
 	TP_fast_assign(
-		__assign_str(name, wq->wchan.name);
+		__assign_str(name, wchan->name);
 	),
 
-	TP_printk("wq=%s", __get_str(name))
+	TP_printk("wchan=%s", __get_str(name))
 );
 
 DECLARE_EVENT_CLASS(mutex_event,
@@ -728,24 +728,24 @@ TRACE_EVENT(evl_timer_shot,
 		  __entry->cycles)
 );
 
-DEFINE_EVENT(wq_event, evl_wait,
-	TP_PROTO(struct evl_wait_queue *wq),
-	TP_ARGS(wq)
+DEFINE_EVENT(wchan_event, evl_wait,
+	TP_PROTO(struct evl_wait_channel *wchan),
+	TP_ARGS(wchan)
 );
 
-DEFINE_EVENT(wq_event, evl_wake_up,
-	TP_PROTO(struct evl_wait_queue *wq),
-	TP_ARGS(wq)
+DEFINE_EVENT(wchan_event, evl_wake_up,
+	TP_PROTO(struct evl_wait_channel *wchan),
+	TP_ARGS(wchan)
 );
 
-DEFINE_EVENT(wq_event, evl_flush_wait,
-	TP_PROTO(struct evl_wait_queue *wq),
-	TP_ARGS(wq)
+DEFINE_EVENT(wchan_event, evl_flush_wait,
+	TP_PROTO(struct evl_wait_channel *wchan),
+	TP_ARGS(wchan)
 );
 
-DEFINE_EVENT(wq_event, evl_finish_wait,
-	TP_PROTO(struct evl_wait_queue *wq),
-	TP_ARGS(wq)
+DEFINE_EVENT(wchan_event, evl_finish_wait,
+	TP_PROTO(struct evl_wait_channel *wchan),
+	TP_ARGS(wchan)
 );
 
 DEFINE_EVENT(mutex_event, evl_mutex_trylock,
diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index abbecb828430..5432694f7c72 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -551,39 +551,6 @@ int evl_trylock_mutex(struct evl_mutex *mutex)
 }
 EXPORT_SYMBOL_GPL(evl_trylock_mutex);
 
-static int wait_mutex_schedule(struct evl_mutex *mutex)
-{
-	struct evl_thread *curr = evl_current();
-	unsigned long flags;
-	int ret = 0, info;
-
-	evl_schedule();
-
-	info = curr->info;
-	if (info & T_RMID)
-		return -EIDRM;
-
-	if (info & (T_TIMEO|T_BREAK)) {
-		raw_spin_lock_irqsave(&mutex->wchan.lock, flags);
-		if (!list_empty(&curr->wait_next)) {
-			list_del_init(&curr->wait_next);
-			if (info & T_TIMEO)
-				ret = -ETIMEDOUT;
-			else if (info & T_BREAK)
-				ret = -EINTR;
-		}
-		raw_spin_unlock_irqrestore(&mutex->wchan.lock, flags);
-	} else if (IS_ENABLED(CONFIG_EVL_DEBUG_CORE)) {
-		bool empty;
-		raw_spin_lock_irqsave(&mutex->wchan.lock, flags);
-		empty = list_empty(&curr->wait_next);
-		raw_spin_unlock_irqrestore(&mutex->wchan.lock, flags);
-		EVL_WARN_ON_ONCE(CORE, !empty);
-	}
-
-	return ret;
-}
-
 /*
  * Undo a PI chain walk due to a locking abort. The mutex state might
  * have changed under us since we dropped mutex->wchan.lock during the
@@ -780,7 +747,7 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 	list_add_priff(curr, &mutex->wchan.wait_list, wprio, wait_next);
 	evl_sleep_on(timeout, timeout_mode, mutex->clock, &mutex->wchan);
 	raw_spin_unlock_irqrestore(&mutex->wchan.lock, flags);
-	ret = wait_mutex_schedule(mutex);
+	ret = __evl_wait_schedule(&mutex->wchan);
 	/* If something went wrong while sleeping, bail out. */
 	if (ret) {
 		raw_spin_lock_irqsave(&mutex->wchan.lock, flags);
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index 486ae38de5a1..f135472e73a5 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -70,7 +70,7 @@ static void __evl_add_wait_queue(struct evl_thread *curr,
 {
 	assert_hard_lock(&wq->wchan.lock);
 
-	trace_evl_wait(wq);
+	trace_evl_wait(&wq->wchan);
 
 	if (!(wq->flags & EVL_WAIT_PRIO))
 		list_add_tail(&curr->wait_next, &wq->wchan.wait_list);
@@ -109,7 +109,7 @@ struct evl_thread *evl_wake_up(struct evl_wait_queue *wq,
 {
 	assert_hard_lock(&wq->wchan.lock);
 
-	trace_evl_wake_up(wq);
+	trace_evl_wake_up(&wq->wchan);
 
 	if (list_empty(&wq->wchan.wait_list)) {
 		waiter = NULL;
@@ -132,7 +132,7 @@ void evl_flush_wait_locked(struct evl_wait_queue *wq, int reason)
 
 	assert_hard_lock(&wq->wchan.lock);
 
-	trace_evl_flush_wait(wq);
+	trace_evl_flush_wait(&wq->wchan);
 
 	list_for_each_entry_safe(waiter, tmp, &wq->wchan.wait_list, wait_next) {
 		list_del_init(&waiter->wait_next);
@@ -172,7 +172,7 @@ void evl_requeue_wait(struct evl_wait_channel *wchan, struct evl_thread *waiter)
 }
 EXPORT_SYMBOL_GPL(evl_requeue_wait);
 
-int evl_wait_schedule(struct evl_wait_queue *wq)
+int __evl_wait_schedule(struct evl_wait_channel *wchan)
 {
 	struct evl_thread *curr = evl_current();
 	unsigned long flags;
@@ -180,7 +180,7 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 
 	evl_schedule();
 
-	trace_evl_finish_wait(wq);
+	trace_evl_finish_wait(wchan);
 
 	/*
 	 * Upon return from schedule, we may or may not have been
@@ -225,7 +225,7 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 		return -ENOMEM;
 
 	if (info & (T_TIMEO|T_BREAK)) {
-		raw_spin_lock_irqsave(&wq->wchan.lock, flags);
+		raw_spin_lock_irqsave(&wchan->lock, flags);
 		if (!list_empty(&curr->wait_next)) {
 			list_del_init(&curr->wait_next);
 			if (info & T_TIMEO)
@@ -233,18 +233,18 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 			else if (info & T_BREAK)
 				ret = -EINTR;
 		}
-		raw_spin_unlock_irqrestore(&wq->wchan.lock, flags);
+		raw_spin_unlock_irqrestore(&wchan->lock, flags);
 	} else if (IS_ENABLED(CONFIG_EVL_DEBUG_CORE)) {
 		bool empty;
-		raw_spin_lock_irqsave(&wq->wchan.lock, flags);
+		raw_spin_lock_irqsave(&wchan->lock, flags);
 		empty = list_empty(&curr->wait_next);
-		raw_spin_unlock_irqrestore(&wq->wchan.lock, flags);
+		raw_spin_unlock_irqrestore(&wchan->lock, flags);
 		EVL_WARN_ON_ONCE(CORE, !empty);
 	}
 
 	return ret;
 }
-EXPORT_SYMBOL_GPL(evl_wait_schedule);
+EXPORT_SYMBOL_GPL(__evl_wait_schedule);
 
 struct evl_thread *evl_wait_head(struct evl_wait_queue *wq)
 {
-- 
2.30.2

